snuba/datasets/storages/validator.py

```diff
@@ -0,0 +1,37 @@
+from __future__ import annotations
+
+from typing import Union
+
+from snuba.datasets.storage import ReadableTableStorage, WritableTableStorage
+from snuba.migrations.groups import get_group_readiness_state_from_storage_set
+
+
+class StorageValidator:
+    def __init__(
+        self, storage: Union[ReadableTableStorage, WritableTableStorage]
+    ) -> None:
+        self.storage = storage
+
+    def validate(self) -> None:
+        self.check_readiness_state_level_with_migration_group()
+
+    def check_readiness_state_level_with_migration_group(self) -> None:
+        """
+        Checks that the storage's readiness state is not greater than the migration group's readiness state.
+        For example, there is no case where a storage is `complete` and the associated migration group is in `limited`.
+        A storage should not be able to query a non-existing table in ClickHouse.
+        This check ensures that the migrations are already deployed before the storage is even loaded.
+        """
+        storage_set_key = self.storage.get_storage_set_key()
+        storage_readiness_state = self.storage.get_readiness_state()
+        group_readiness_state = get_group_readiness_state_from_storage_set(
+            storage_set_key
+        )
+        if storage_readiness_state.level > group_readiness_state.level:
```

Comment:
Was the Enum `__new__` override to support this check? You could also add a `__gt__` operator override I think, which would then mean you don't need the level parameter.

Reply:
I like the `__gt__` override better. Updated.